File: 0155-min-stack/0155-min-stack.cpp
class MinStack {
public:
    stack<pair<int,int>> a1;
    MinStack() {
        
    }
    
    void push(int val) {
        if(a1.empty()==true){
            a1.push(make_pair(val,val));
        }
        else
            a1.push(make_pair(val,min(a1.top().second,val)));
    }
    
    void pop() {
        a1.pop();
    }
    
    int top() {
       return a1.top().first;
    }
    
    int getMin() {
        return a1.top().second;
    }
};

/**
 * Your MinStack object will be instantiated and called as such:
 * MinStack* obj = new MinStack();
 * obj->push(val);
 * obj->pop();
 * int param_3 = obj->top();
 * int param_4 = obj->getMin();
 */

File: 0155-min-stack/0155-min-stack.java
class MinStack {
    Stack<Integer> stack;
    Stack<Integer> minS;

    public MinStack() {
        stack = new Stack<>();
        minS = new Stack<>();
    }

    public void push(int x) {

        if (stack.isEmpty()) {
            stack.push(x);
            minS.push(x);
        } else {
            stack.push(x);
            if (x <= minS.peek()) minS.push(x);
        }
    }

    public void pop() {
        int sp = stack.peek();
        int mp = minS.peek();
        if (sp == mp) {
            stack.pop();
            minS.pop();
        } else {
            stack.pop();
        }


    }

    public int top() {
        return stack.peek();
    }

    public int getMin() {
        return minS.peek();
    }
}



